### PR TITLE
Add counterfactual minus guided/unguided diff at location level

### PR DESCRIPTION
**Pull request by Zapiano**
Followed an approach similar to https://github.com/open-AIMS/ADRIA.jl/issues/452#issuecomment-1942888428 but for a spatial level.

Note: Depends on #847 

## Example of code

```julia
using ADRIA
using WGLMakie, GeoMakie, GraphMakie

Makie.inline!(false)

# Load domain
dom = ADRIA.load_domain(...)

# Create scenarios
num_scens = 2^10
scens = ADRIA.sample(dom, num_scens)

# Run model / Load ResultSet
rs = ADRIA.run_scenarios(dom, scens, "45")

# Extract cf_difference_loc for guided and unguided (in that order)
rc = ADRIA.metrics.relative_cover(rs)

# Compute difference between guided and counterfactual using the 0.6-th quantile
gd_res = metrics.ensemble_loc_difference(rc, scens; agg_metric=0.6)

# Compute difference between unguided and counterfactual using the median
ug_res = metrics.ensemble_loc_difference(r_cover, scens; diff_target=:unguided)

# Plot maps of difference to the counterfactual
ADRIA.viz.map(rs, gd_res[summary=At(:agg_value)]; diverging=true)
ADRIA.viz.map(rs, ug_res[summary=At(:agg_value)]; diverging=true)
```

## Plots
- 1024 scenarios

| Guided - Counterfactual (0.6-th quantile) | Unguided - Counterfactual (median)|
| ----------------------- | ------------------------- |
| gd_cf_diff_map | ug_cf_diff_map |

### CODE REVIEW

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [inline src/metrics/spatial.jl:4]: High-level docstrings should be at the very top of file.

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [inline src/metrics/spatial.jl:143]: Could reuse the first `cf_shuf` here

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [inline src/metrics/spatial.jl:116]:
```suggestion
Give the mean bootstrapped difference from the counterfactual for a given metric.
```

Comment (ConnectedSystems) [reply src/metrics/spatial.jl:116]: Document what is returned?

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [inline src/metrics/spatial.jl:134]: Not interested in the confidence bounds?

**Comment by ConnectedSystems** [comment]
Bump @Zapiano

**Comment by ConnectedSystems** [comment]
How close are we to merging this?

**Comment by Zapiano** [comment]
> How close are we to merging this?

I'll start looking into this now and let you know what's missing or of it's ready soon.

**Comment by Zapiano** [comment]
@ConnectedSystems This one is ready for another review.

**Comment by ConnectedSystems** [comment]
An example usage would be appreciated too @Zapiano

**Comment by Zapiano** [comment]
> Thanks.
> 
> An example usage would be appreciated too @Zapiano

I've added an example in this PR. Once this PR is approved I can add this plot to the documentation on a separate PR.

**Comment by ConnectedSystems** [comment]
I'm not sure the indicated use it the best approach.

It forces only one set of bootstrapped results (only the median?), and forces additional computations to be done (e.g., what if I only want `gd_res`?)

```julia
# Extract cf_difference_loc for guided and unguided (in that order)
rc = ADRIA.metrics.relative_cover(rs)
gd_res, ug_res = ADRIA.metrics.cf_difference_loc(rc, scens)  # <- this method name feels awkward

# Plot cf_difference_loc for unguided and guided
gd_fig = ADRIA.viz.diff_map(rs, gd_res[2, :])  # <- why `[2, :]`? (I know why, but feels awkward)
ug_fig = ADRIA.viz.diff_map(rs, ug_res[2, :])
```

Instead, I suggest something like the below which, to me, is a cleaner UX.

```julia
rc = ADRIA.metrics.relative_cover(rs)

# outcome, scenario spec, percentile value of interest
gd_res = ADRIA.metrics.ensemble_difference(rc, scens, 0.5; timestep=2, scenario_type=:guided)

# outcome, scenario spec, statistical function of interest
ug_res = ADRIA.metrics.ensemble_difference(rc, scens, mean; timestep=2, scenario_type=:unguided)
```

On the plotting, it kind of feels like we're getting to the point where we want our own YAXArray subtype. 

```julia
ADRIA.viz.map(rs, gd_res::SomeCustomYAXArray?)  #<- uses different defaults to the "usual" `viz.map()`
```

**Comment by Zapiano** [comment]
@ConnectedSystems this was updated to address your comments. The description was updated with new example of usage and plots.

**Comment by Zapiano** [comment]
> On the plotting, it kind of feels like we're getting to the point where we want our own YAXArray subtype.
> 
> ```julia
> ADRIA.viz.map(rs, gd_res::SomeCustomYAXArray?)  #<- uses different defaults to the "usual" `viz.map()`
> ```

I'm not sure I understand the purpose of having our own YAXArray subtype.

**Comment by ConnectedSystems** [comment]
> I'm not sure I understand the purpose of having our own YAXArray subtype.

To be able to define our own custom styling for specific use cases and be able to make different styles when plotting some arbitrary results (e.g., `some_plot(a_matrix)`) compared to `some_plot(a_general_yaxarray)` and `some_plot(a_yaxarray_with_context_specific_metadata)`

**Comment by ConnectedSystems** [comment]
@Zapiano fyi there are conflicts now because of the merge of #863. Could you rebase or merge main as appropriate?

**Comment by Zapiano** [comment]
> @Zapiano fyi there are conflicts now because of the merge of #863. Could you rebase or merge main as appropriate?

I rebased but the tests are not passing. I'll let you know when this is ready again.

**Comment by Zapiano** [comment]
> > @Zapiano fyi there are conflicts now because of the merge of #863. Could you rebase or merge main as appropriate?
> 
> I rebased but the tests are not passing. I'll let you know when this is ready again.

@ConnectedSystems  the tests that were not passing is a problem on the main branch apparently (https://github.com/open-AIMS/ADRIA.jl/issues/890). If you prefer waiting for that to be fixed so we are sure the tests are passing for this PR as well it's fine, otherwise this one is ready to merge.

**Comment by ConnectedSystems** [comment]
Have you updated to the latest dependencies?

**Comment by Zapiano** [comment]
> Have you updated to the latest dependencies?

Yes.

**Comment by ConnectedSystems** [comment]
> I'm not sure I understand the purpose of having our own YAXArray subtype.

FYI, Makie's theme recipes seem more stable now, and better documented:

https://docs.makie.org/stable/tutorials/wrap-existing-recipe

**Comment by ConnectedSystems** [comment]
This PR can be merged once #890 is confirmed to be addressed by #891

**Comment by Zapiano** [comment]
@ConnectedSystems tests are passing and I've updated testset names for the tests that are related to this PR as we discussed.

**Comment by Zapiano** [comment]
@ConnectedSystems just reminding you that this is ready for review.

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [inline src/metrics/spatial.jl:165]: Incorrect example here?

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [inline src/metrics/spatial.jl:167]:
```suggestion
# Plot maps of difference to the counterfactual at timestep 2
```

Is the above correct?

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [inline src/metrics/spatial.jl:181]: Minor point - could we use a global random seed instead?

```suggestion
    rng::AbstractRNG=Random.GLOBAL_RNG
```

Later on...

`shuffle(rng, 1:n_cf_outcomes)[1:min_n_outcomes]`

File: src/metrics/spatial.jl
Comment (Zapiano) [reply src/metrics/spatial.jl:167]: Yes, this `2` is not timestep, `gd_res[1,:]` and `gd_res[3,:]` are the lower and upper bounds of the confidence interval and `gd_res[2,:]` is the metric used (median or whatever).

File: src/metrics/spatial.jl
Comment (Zapiano) [reply src/metrics/spatial.jl]: Fixed.

File: src/metrics/spatial.jl
Comment (Zapiano) [reply src/metrics/spatial.jl:181]: Does that mean that the user shouldn't be allowed to set the seed and I should also remove this `Random.seed!(rand_seed)` ?

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Two minor comments, but this is ready to be merged once those are addressed.

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [reply src/metrics/spatial.jl]: Yes, you should remove the `Random.seed!` inside the function.

It means the user can set it outside the function.

This way the state of the random number generator does not "leak" outside the scope of the function.
Currently, the function "resets" the seed state, which may affect other functions.

Best practice is to set a RNG and pass that to each function if necessary

```julia
rng = Random.Xoshiro(123)  # use a different RNG algorithm
some_function(args...; rng=rng)
another_function(args...; rng=rng)
```

As an aside: the current exception in ADRIA is the scenario running, where I made the decision that we want the outcome to be tied to the scenario factors. But it's implemented in a suboptimal way... another thing to revisit.

Comment (Zapiano) [reply src/metrics/spatial.jl]: Makes sense, thanks for the explanation.

**Comment by Zapiano** [comment]
> Two minor comments, but this is ready to be merged once those are addressed.

Done, and all tests passing.

File: src/metrics/spatial.jl
Comment (ConnectedSystems) [reply src/metrics/spatial.jl:167]: Hmm, is `gd_res` a YAXArray? It would be clearer what the "2" is for the example if the axis name is used as the selector instead

Comment (ConnectedSystems) [reply src/metrics/spatial.jl:167]: I'd ask that the comment be updated to clarify what the "2" is.

**Comment by Zapiano** [comment]
@ConnectedSystems I made two changes:

1. I updated the example (and tests) to use the YAXArray dim/label names instead of the index `2` directly when assessing the aggregate metric 
2. I have deleted the function `diff_map` and I'm using `map` instead, with an option for `diverging` color palette. That seems more flexible if we want to plot other maps that also need this palette in the future but are not "diff maps".

**Pull request merged**
